File: hanoi/src/Hanoi.java
import java.util.Arrays;
import java.util.Stack;

public class Hanoi {

    Stack<Integer> sourcePeg = new Stack<Integer>();
    Stack<Integer> targetPeg = new Stack<Integer>();
    Stack<Integer> helperPeg = new Stack<Integer>();

    private void solve(int disks) {
        for (int i = disks; i > 0; i--) {
            sourcePeg.push(i);
        }
        System.out.println("Playing Hanoi for " + disks + " disks:\n");
        showPegs();
        move(disks, sourcePeg, targetPeg, helperPeg);
        System.out.println("Solved.");
    }

    private void move(int n, Stack<Integer> sourcePeg, Stack<Integer> targetPeg, Stack<Integer> helperPeg) {
        if (n > 0) {
            if (n > 1) {
                move(n - 1, sourcePeg, helperPeg, targetPeg);
            }
            targetPeg.push(sourcePeg.pop());
            showPegs();
            if (n > 1) {
                move(n - 1, helperPeg, targetPeg, sourcePeg);
            }
        }
    }

    private void showPegs() {
        System.out.println("sourcePeg = " + Arrays.toString(sourcePeg.toArray()));
        System.out.println("helperPeg = " + Arrays.toString(helperPeg.toArray()));
        System.out.println("targetPeg = " + Arrays.toString(targetPeg.toArray()) + "\n");
    }

    public static void main(String[] args) {
        Hanoi hanoi = new Hanoi();
        hanoi.solve(5);
    }
}

File: python_challenges/hanoi.py
__author__ = 'tilmann.bruckhaus'


class Hanoi:

    def __init__(self, disks):
        print "Playing \"Towers of Hanoi\" for ", disks, " disks:"
        self.disks = disks
        self.source_peg = []
        self.helper_peg = []
        self.target_peg = []
        self.set_up_pegs()
        self.show()

    def solve(self):
        self.move(self.disks, self.source_peg, self.target_peg, self.helper_peg)
        print "\nSolved."

    def move(self, peg_count, source_peg, target_peg, helper_peg):
        if peg_count >= 1:
            self.move(peg_count - 1, source_peg, helper_peg, target_peg)
            target_peg.append(source_peg.pop())
            self.show()
            self.move(peg_count - 1, helper_peg, target_peg, source_peg)

    def show(self):
        print "\nsource_pge: ", self.source_peg
        print "helper_pge: ", self.helper_peg
        print "target_pge: ", self.target_peg

    def set_up_pegs(self):
        for i in reversed(range(1, self.disks + 1)):
            self.source_peg.append(i)

if __name__ == '__main__':
    Hanoi(4).solve()